File: dynamic-programming/optimal-binary-search-tree/python/top_down_optimal_binary_search_tree.py
import sys


def optimal_bst(p, q, i, j):
    results = [[sys.maxsize for _ in range(
        max(len(p), len(q) + 1))] for _ in range(max(len(p), len(q) + 1))]

    return optimal_bst_internal(p, q, i, j, results)


def optimal_bst_internal(p, q, i, j, results):
    if results[i][j] < sys.maxsize:
        return results[i][j]

    if j == i - 1:
        result = q[i - 1]
    else:
        w = q[i - 1]
        for l in range(i, j + 1):
            w = w + p[l] + q[l]
        w = round(w, 2)

        result = sys.maxsize
        for r in range(i, j + 1):
            result = min(result, optimal_bst_internal(
                p, q, i, r - 1, results) + optimal_bst_internal(p, q, r + 1, j, results) + w)

    results[i][j] = result

    return result


if __name__ == '__main__':
    p = [0.0, 0.15, 0.10, 0.05, 0.10, 0.20]
    q = [0.05, 0.10, 0.05, 0.05, 0.05, 0.10]

    print("  i |", end='')
    for i in range(len(p)):
        print(f"{i:^5}|", end='')
    print("")
    print("p[i]|", end='')
    for _p in p:
        print(f"{_p:>5}|", end='')
    print("")
    print("q[i]|", end='')
    for _q in q:
        print(f"{_q:>5}|", end='')
    print("")

    diff = -1
    while diff < len(p) - 1:
        print(f"--- diff = {diff} ---")

        i = 1
        j = i + diff
        while j < len(p):
            result = optimal_bst(p, q, i, j)

            print(f"i = {i}, j = {j} / result = {result:.2f}")

            i += 1
            j += 1

        diff += 1

File: dynamic-programming/optimal-binary-search-tree/golang/top_down_optimal_binary_search_tree.go
package main

import (
	"fmt"
	"math"
)

func optimalBst(p, q []float64, i, j int) float64 {
	results := make([][]float64, len(p)+1)

	for k := 0; k < len(results); k++ {
		results[k] = make([]float64, len(p)+1)
	}

	return optimalBstInternal(p, q, i, j, results)
}

func optimalBstInternal(p, q []float64, i, j int, results [][]float64) float64 {
	if results[i][j] > 0.0 {
		return results[i][j]
	}

	var result float64
	if j == i-1 {
		result = q[i-1]
	} else {
		w := q[j-1]
		for l := i; l <= j; l++ {
			w = w + p[l] + q[l]
		}

		result = math.MaxFloat64
		for r := i; r <= j; r++ {
			result = min(result, optimalBstInternal(p, q, i, r-1, results)+optimalBstInternal(p, q, r+1, j, results)+w)
		}
	}

	results[i][j] = result

	return result
}

func main() {
	p := []float64{0.0, 0.15, 0.10, 0.05, 0.10, 0.20}
	q := []float64{0.05, 0.10, 0.05, 0.05, 0.05, 0.10}

	fmt.Print("  i |")
	for i := range p {
		fmt.Printf("%5d|", i)
	}
	fmt.Print("\n")
	fmt.Print("p[i]|")
	for _, _p := range p {
		fmt.Printf("%5.2f|", _p)
	}
	fmt.Print("\n")
	fmt.Print("q[i]|")
	for _, _q := range q {
		fmt.Printf("%5.2f|", _q)
	}
	fmt.Print("\n")

	for diff := -1; diff < len(p)-1; diff++ {
		fmt.Printf("--- diff = %d ---\n", diff)

		i := 1
		for j := i + diff; j < len(p); j++ {
			result := optimalBst(p, q, i, j)

			fmt.Printf("i = %d, j = %d / result = %.2f\n", i, j, result)

			i++
		}
	}
}
